rtnetlink/src/ns.rs

```diff
@@ -0,0 +1,223 @@
+                log::trace!("Child creating namespace");
+
+                let mut netns_path = String::new();
+
+                let dir_path = Path::new(NETNS_PATH);
+                let mut mkdir_mode = Mode::empty();
+                let mut open_flags = OFlag::empty();
+                let mut mount_flags = nix::mount::MsFlags::empty();
+                let mut setns_flags = CloneFlags::empty();
+                let none_fs = Path::new(&NONE_FS);
+                let none_p4: Option<&Path> = None;
+
+                // flags in mkdir
+                mkdir_mode.insert(Mode::S_IRWXU);
+                mkdir_mode.insert(Mode::S_IRGRP);
+                mkdir_mode.insert(Mode::S_IXGRP);
+                mkdir_mode.insert(Mode::S_IROTH);
+                mkdir_mode.insert(Mode::S_IXOTH);
+
+                open_flags.insert(OFlag::O_RDONLY);
+                open_flags.insert(OFlag::O_CREAT);
+                open_flags.insert(OFlag::O_EXCL);
+
+                netns_path.push_str(NETNS_PATH);
+                netns_path.push_str(&ns_name);
+
+                // creating namespaces folder if not exists
+                #[allow(clippy::collapsible_if)]
+                if nix::sys::stat::stat(dir_path).is_err() {
+                    if nix::unistd::mkdir(dir_path, mkdir_mode).is_err() {
+                        exit(-1);
+                    }
+                }
+
+                mount_flags.insert(nix::mount::MsFlags::MS_BIND);
+                mount_flags.insert(nix::mount::MsFlags::MS_REC);
+
+                if nix::mount::mount(
+                    Some(Path::new(dir_path)),
+                    dir_path,
+                    Some(none_fs),
+                    mount_flags,
+                    none_p4,
+                )
+                .is_err()
+                {
+                    exit(-1);
+                }
+
+                mount_flags = nix::mount::MsFlags::empty();
+                mount_flags.insert(nix::mount::MsFlags::MS_SHARED);
+                mount_flags.insert(nix::mount::MsFlags::MS_REC);
+                if nix::mount::mount(
+                    Some(Path::new("")),
+                    dir_path,
+                    Some(none_fs),
+                    mount_flags,
+                    none_p4,
+                )
+                .is_err()
+                {
+                    exit(-1);
+                }
+
+                let ns_path = Path::new(&netns_path);
+
+                // creating the netns file
+                let fd = match nix::fcntl::open(ns_path, open_flags, Mode::empty()) {
+                    Ok(raw_fd) => raw_fd,
+                    Err(_) => exit(-1),
+                };
+
+                if nix::unistd::close(fd).is_err() {
+                    let _ = nix::unistd::unlink(ns_path);
+                    exit(-1)
+                }
+
+                // unshare to the new network namespace
+                if nix::sched::unshare(CloneFlags::CLONE_NEWNET).is_err() {
```

Comment:
If I understand correctly, in multithreaded environments, `unshare` has effect on the _current thread_, not the current process. So I wonder if what we're doing is correct.

Reply:
This was the real tricky part, indeed only the thread calling unshare does actually `unshare` itself from the parent environment, actually, I was also afraid to use the `fork` because as we are using tokio it is not really safe to fork a multithreaded process, it may hang because of a lock taken by the `parent` and that the `child` never see as released.

That said, I'm calling `unshare` in the child process, which may still be multithreaded because it starts as part of a tokio runtime.
So I do not really see other options that make this a sync interface. 
I'm really open to suggestions

Reply:
I implemented code, and I avoided starting tokio in the parent or child until the fork() and unshare(). 
Typical users will want to do something in the child (in the new namespace), so I don't think that calling fork(2) is the right thing. I think rather, all of the rest of the stuff (the mkdirs and whatnot) should be packaged into functions (not async ones), and document how to use this.

Reply:
The current code is actually mimicking the `ip netns add` which does not give the possibility to the user to execute something in the new namespace, that one is the `ip netns exec NS_NAME CMD`, in that case, I would agree with you that we should leave the possibility to the user to do something in the new namespace.

I really need this for https://github.com/eclipse-fog05/fog05/pull/208 and as I already tested across various distros and CPU architecture I would tend to just merge it.

Reply:
While "ip netns add" is good to have, if we just wanted to run two scripts, we wouldn't be doing this in Rust, right? :-)
I would prefer to have all the before-fork(2), between-fork(2)-and-unshare(2), and any after-unshare(2) code refactored so that people can call it appropriately.  The result would be:

````
     NetworkNamespace::prep_for_fork(2);
     match unsafe { fork() } {
            Ok(ForkResult::Parent { child, .. }) => { NetworkNamespace::parent_processing(child); }
            Ok(ForkResult::Child) => { NetworkNamespace::child_process();  
                                                       SomeOtherProcessing(); 
                                                       NetworkNamespace::unshare_processing(); }
    }

~~~~

Will we support setns() properly?  There all sorts of interactions with tokio threads which need to be dealt with.

Reply:
Sure I can refactor the code, but I still think that provide something like `ip netns exec NS CMD` is very difficult because only the thread calling the `unshare`+`setns`+some `mount` will be moved to the specified namespace.

Let me just explain how we use this in [fog05-linux-networking](https://github.com/eclipse-fog05/fog05/blob/master/linux-network/src/networking.rs#L926)

We have "something" asking this component to create a new network namespace, here we call the code from `netlink`, then we spawn a separate process to "manage" this new namespace (e.g. create new interfaces, nftables, etc.).

Anyway, I will refactor the code as you suggest it will also improve readability.